### ISSUE TITLE
HttpClient SocketsHttpHandler ignores SslClientAuthenticationOptions

### ISSUE THREAD

**Issue by tick-rick** [dotnet/corefx#42670]
Moving from https://github.com/dotnet/corefx/issues/21577#issuecomment-556932989
Package installed via (dotnet add package System.Net.Http --version 4.3.4)

I think I must be missing something with that API. The following code enables 1 cipher suite, yet the result in wireshark TLS1.3 ClientHello lists 31 cipher suites (of which TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 is indeed one, but that is beside the question)

            var sslOptions = new SslClientAuthenticationOptions {
                CipherSuitesPolicy  = new CipherSuitesPolicy(new List<TlsCipherSuite>{ TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 })
            };
            var socketsHttpHandler = new SocketsHttpHandler { SslOptions = sslOptions };
            var httpClient = new HttpClient(socketsHttpHandler, true);
            httpClient.GetAsync("https://www.google.com").Wait();

## Dotnet Info
```
$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100
 Commit:    04339c3a26

Runtime Environment:
 OS Name:     debian
 OS Version:
 OS Platform: Linux
 RID:         debian-x64
 Base Path:   /home/user/.dotnet/sdk/3.0.100/

Host (useful for support):
  Version: 3.0.0
  Commit:  7d57652f33

.NET Core SDKs installed:
  2.1.802 [/home/user/.dotnet/sdk]
  3.0.100-rc1-014190 [/home/user/.dotnet/sdk]
  3.0.100 [/home/user/.dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [/home/user/.dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [/home/user/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-rc1.19457.4 [/home/user/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [/home/user/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [/home/user/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-rc1-19456-20 [/home/user/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [/home/user/.dotnet/shared/Microsoft.NETCore.App]
```
To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

**Comment by krwq** [comment]
@tick-rick can you try running OpenSSL server:

```shell
# generate cert
openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem -days 365 -nodes
# run server
openssl s_server -key key.pem -cert cert.pem -accept 44330 -www
```

and following code:

```csharp
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Security;

namespace Repro
{
    class Program
    {
        static void Main(string[] args)
        {
            var sslOptions = new SslClientAuthenticationOptions
            {
                CipherSuitesPolicy = new CipherSuitesPolicy(new List<TlsCipherSuite> { TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 }),
                RemoteCertificateValidationCallback = (a, b, c, d) => true,
            };

            var socketsHttpHandler = new SocketsHttpHandler { SslOptions = sslOptions };
            var httpClient = new HttpClient(socketsHttpHandler, true);
            HttpResponseMessage resp = httpClient.GetAsync("https://localhost:44330").Result;
            Console.WriteLine(resp.Content.ReadAsStringAsync().Result);
        }
    }
}
```

and print the output?

also please provide output of `openssl version`

I've tried running above on Ubuntu and I correctly get a single ciphersuite

**Comment by tick-rick** [comment]
OpenSSL version: OpenSSL 1.1.1d  10 Sep 2019


results of dotnet run for the above:
```
<HTML><BODY BGCOLOR="#ffffff">
<pre>

s_server -key key.pem -cert cert.pem -accept 44330 -www 
Secure Renegotiation IS supported
Ciphers supported in s_server binary
TLSv1.3    :TLS_AES_256_GCM_SHA384    TLSv1.3    :TLS_CHACHA20_POLY1305_SHA256 
TLSv1.3    :TLS_AES_128_GCM_SHA256    TLSv1.2    :ECDHE-ECDSA-AES256-GCM-SHA384 
TLSv1.2    :ECDHE-RSA-AES256-GCM-SHA384 TLSv1.2    :DHE-RSA-AES256-GCM-SHA384 
TLSv1.2    :ECDHE-ECDSA-CHACHA20-POLY1305 TLSv1.2    :ECDHE-RSA-CHACHA20-POLY1305 
TLSv1.2    :DHE-RSA-CHACHA20-POLY1305 TLSv1.2    :ECDHE-ECDSA-AES128-GCM-SHA256 
TLSv1.2    :ECDHE-RSA-AES128-GCM-SHA256 TLSv1.2    :DHE-RSA-AES128-GCM-SHA256 
TLSv1.2    :ECDHE-ECDSA-AES256-SHA384 TLSv1.2    :ECDHE-RSA-AES256-SHA384   
TLSv1.2    :DHE-RSA-AES256-SHA256     TLSv1.2    :ECDHE-ECDSA-AES128-SHA256 
TLSv1.2    :ECDHE-RSA-AES128-SHA256   TLSv1.2    :DHE-RSA-AES128-SHA256     
TLSv1.0    :ECDHE-ECDSA-AES256-SHA    TLSv1.0    :ECDHE-RSA-AES256-SHA      
SSLv3      :DHE-RSA-AES256-SHA        TLSv1.0    :ECDHE-ECDSA-AES128-SHA    
TLSv1.0    :ECDHE-RSA-AES128-SHA      SSLv3      :DHE-RSA-AES128-SHA        
TLSv1.2    :RSA-PSK-AES256-GCM-SHA384 TLSv1.2    :DHE-PSK-AES256-GCM-SHA384 
TLSv1.2    :RSA-PSK-CHACHA20-POLY1305 TLSv1.2    :DHE-PSK-CHACHA20-POLY1305 
TLSv1.2    :ECDHE-PSK-CHACHA20-POLY1305 TLSv1.2    :AES256-GCM-SHA384         
TLSv1.2    :PSK-AES256-GCM-SHA384     TLSv1.2    :PSK-CHACHA20-POLY1305     
TLSv1.2    :RSA-PSK-AES128-GCM-SHA256 TLSv1.2    :DHE-PSK-AES128-GCM-SHA256 
TLSv1.2    :AES128-GCM-SHA256         TLSv1.2    :PSK-AES128-GCM-SHA256     
TLSv1.2    :AES256-SHA256             TLSv1.2    :AES128-SHA256             
TLSv1.0    :ECDHE-PSK-AES256-CBC-SHA384 TLSv1.0    :ECDHE-PSK-AES256-CBC-SHA  
SSLv3      :SRP-RSA-AES-256-CBC-SHA   SSLv3      :SRP-AES-256-CBC-SHA       
TLSv1.0    :RSA-PSK-AES256-CBC-SHA384 TLSv1.0    :DHE-PSK-AES256-CBC-SHA384 
SSLv3      :RSA-PSK-AES256-CBC-SHA    SSLv3      :DHE-PSK-AES256-CBC-SHA    
SSLv3      :AES256-SHA                TLSv1.0    :PSK-AES256-CBC-SHA384     
SSLv3      :PSK-AES256-CBC-SHA        TLSv1.0    :ECDHE-PSK-AES128-CBC-SHA256 
TLSv1.0    :ECDHE-PSK-AES128-CBC-SHA  SSLv3      :SRP-RSA-AES-128-CBC-SHA   
SSLv3      :SRP-AES-128-CBC-SHA       TLSv1.0    :RSA-PSK-AES128-CBC-SHA256 
TLSv1.0    :DHE-PSK-AES128-CBC-SHA256 SSLv3      :RSA-PSK-AES128-CBC-SHA    
SSLv3      :DHE-PSK-AES128-CBC-SHA    SSLv3      :AES128-SHA                
TLSv1.0    :PSK-AES128-CBC-SHA256     SSLv3      :PSK-AES128-CBC-SHA        
---
Ciphers common between both SSL end points:
TLS_AES_256_GCM_SHA384     TLS_CHACHA20_POLY1305_SHA256 TLS_AES_128_GCM_SHA256    
ECDHE-ECDSA-AES256-GCM-SHA384 ECDHE-RSA-AES256-GCM-SHA384 DHE-RSA-AES256-GCM-SHA384 
ECDHE-ECDSA-CHACHA20-POLY1305 ECDHE-RSA-CHACHA20-POLY1305 DHE-RSA-CHACHA20-POLY1305 
ECDHE-ECDSA-AES128-GCM-SHA256 ECDHE-RSA-AES128-GCM-SHA256 DHE-RSA-AES128-GCM-SHA256 
ECDHE-ECDSA-AES256-SHA384  ECDHE-RSA-AES256-SHA384    DHE-RSA-AES256-SHA256     
ECDHE-ECDSA-AES128-SHA256  ECDHE-RSA-AES128-SHA256    DHE-RSA-AES128-SHA256     
ECDHE-ECDSA-AES256-SHA     ECDHE-RSA-AES256-SHA       DHE-RSA-AES256-SHA        
ECDHE-ECDSA-AES128-SHA     ECDHE-RSA-AES128-SHA       DHE-RSA-AES128-SHA        
AES256-GCM-SHA384          AES128-GCM-SHA256          AES256-SHA256             
AES128-SHA256              AES256-SHA                 AES128-SHA
Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:Ed25519:Ed448:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:ECDSA+SHA1:RSA+SHA224:RSA+SHA1:DSA+SHA224:DSA+SHA1:DSA+SHA256:DSA+SHA384:DSA+SHA512
Shared Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:Ed25519:Ed448:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:ECDSA+SHA1:RSA+SHA224:RSA+SHA1
Supported Elliptic Groups: X25519:P-256:X448:P-521:P-384
Shared Elliptic groups: X25519:P-256:X448:P-521:P-384
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: A21EB787522D2A2626CBB7A2A812284C573A36DCE09C30AF483A7A91BA2E0BB0
    Session-ID-ctx: 01000000
    Resumption PSK: 8AA9F5F64C569F573A42C38CCAAE130D34D24A17DBB8891AD3AE774E6DA464D54810650A6AE4F21842E0CAE07A7EFD01
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1574378796
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
   0 items in the session cache
   0 client connects (SSL_connect())
   0 client renegotiates (SSL_connect())
   0 client connects that finished
   2 server accepts (SSL_accept())
   0 server renegotiates (SSL_accept())
   2 server accepts that finished
   0 session cache hits
   0 session cache misses
   0 session cache timeouts
   0 callback cache hits
   0 cache full overflows (128 allowed)
---
no client certificate available
</pre></BODY></HTML>
```

**Comment by krwq** [comment]
Thanks, I'll investigate this. This looks very fishy

**Comment by tick-rick** [comment]
FWIW, I made a TlsClient in BouncyCastle with that exact cipher in repro and got this on the server:
```
140217379460288:error:1409442F:SSL routines:ssl3_read_bytes:tlsv1 alert insufficient security:../ssl/record/rec_layer_s3.c:1543:SSL alert number 71
```
and this on the client:
```
Unhandled exception. System.Exception: Exception raised by job
 ---> Org.BouncyCastle.Crypto.Tls.TlsFatalAlert: insufficient_security(71)
```

Which makes me wonder if openssl behaviour has changed recently so that if you provide what it thinks is an insecure cipher suite, it uses a default list.

**Comment by scalablecory** [comment]
> Package installed via (dotnet add package System.Net.Http --version 4.3.4)

Unrelated to this issue, please avoid installing via nuget package. HttpClient is provided in-box.

**Comment by tick-rick** [comment]
@scalablecory Ok, noted

**Comment by krwq** [comment]
@tick-rick can you try the same but remove `-www` switch from openssl command and following code (skip HttpClient):

```csharp
            using (TcpClient client = new TcpClient("localhost", 44330))
            {
                using (SslStream ssl = new SslStream(client.GetStream()))
                {
                    var options = new SslClientAuthenticationOptions()
                    {
                        TargetHost = "localhost",
                        CipherSuitesPolicy = new CipherSuitesPolicy(new List<TlsCipherSuite> { TlsCipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 }),
                    };
                    ssl.AuthenticateAsClientAsync(options).Wait();
                }
            }
```

(only interested in output of openssl)

If this looks correct then likely some HttpClient plumbing is ignoring CipherSuites policy

**Comment by tick-rick** [comment]
```
$ openssl s_server -key key.pem -cert cert.pem -accept 44330
Using default temp DH parameters
ACCEPT
-----BEGIN SSL SESSION PARAMETERS-----
MFoCAQECAgMDBAIAawQABDD2wEnyiPm1UfTMo92Dp0MkLl5zb8B2ddVauLWt70HN
ULVOJyXLpDJH7f4096bbf8KhBgIEXdcjpKIEAgIcIKQGBAQBAAAArQMCAQE=
-----END SSL SESSION PARAMETERS-----
Shared ciphers:DHE-RSA-AES256-SHA256
Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:Ed25519:Ed448:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:ECDSA+SHA1:RSA+SHA224:RSA+SHA1:DSA+SHA224:DSA+SHA1:DSA+SHA256:DSA+SHA384:DSA+SHA512
Shared Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:Ed25519:Ed448:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:ECDSA+SHA1:RSA+SHA224:RSA+SHA1:DSA+SHA224:DSA+SHA1:DSA+SHA256:DSA+SHA384:DSA+SHA512
CIPHER is DHE-RSA-AES256-SHA256
Secure Renegotiation IS supported
ERROR
shutting down SSL
CONNECTION CLOSED
```

**Comment by krwq** [comment]
This looks good which seems the SocketsHttpHandler seem to be ignoring ssl options

**Comment by krwq** [comment]
@dotnet/ncl does any of you have cycles too look at this? This seems pretty important to fix

**Comment by karelz** [comment]
@krwq what is the important scenario? Can you please summarize it?

**Comment by wfurt** [comment]
This is similar to #36555. It should not be hard to fix. I think we should take it for next release. (and add more tests)

**Comment by krwq** [comment]
@karelz, security settings don't respect what user asks for (I think you will get defaults instead which are not insecure but they're not what user requested - even exception would be more appropriate).

**Comment by krwq** [comment]
@wfurt, yes they both seem the same to me. Feel free to close one or another as dup

**Comment by stephentoub** [comment]
The problem is here:
https://github.com/dotnet/runtime/blob/be980b71efadc622b5720a36867696758e59e71c/src/libraries/Common/src/System/Net/Security/SslClientAuthenticationOptionsExtensions.cs#L9-L21

When CipherSuitesPolicy was added to SslClientAuthenticationOptions, this wasn't updated with the additional property.

**Comment by geoffkizer** [comment]
@stephentoub Should there be a copy ctor, then? What's the best way to avoid this sort of issue in the future?

**Comment by stephentoub** [comment]
Probably.  For now I added a big honkin' debug block in https://github.com/dotnet/runtime/pull/211.

**Comment by stephentoub** [comment]
I merged https://github.com/dotnet/runtime/pull/211 to address this in 5.0, but re-opening this and moving to 3.1 for @karelz to decide if it should be brought back to 3.1.

**Comment by davidsh** [comment]
@danmosemsft - I think this is something we should put into 3.1. It was an oversight when we added CipherSuitesPolicy. Without this fix, the policy doesn't work for HTTP scenarios.

**Comment by danmosemsft** [comment]
@davidsh let's get a PR and template up

**Comment by stephentoub** [comment]
> let's get a PR and template up

https://github.com/dotnet/corefx/pull/42694

**Comment by karelz** [comment]
Fixed in release/3.1 in PR #42694 and in 5.0 in PR https://github.com/dotnet/runtime/pull/211
